Refactor Exec Parsers - remove individual parser classes

nartal1 opened the pull request:
This PR contributes to #1325 . This is the first iteration to clean up ExecParsers code. 

This pull request refactors several execution parsers in the `planparser` package to use a new generic execution parser class. The changes aim to simplify the codebase by consolidating common logic into a single class, `GenericExecParser`.
In this PR, following changes are done:

1. If the Execs don't have expressions, then execName is assigned and ExecInfo object is created from GenericExecParser.
2. If the Execs have expressions then appropriate expressionParser is passed to the constructor of GenericExexParser and ExecInfo object is created from GenericParser. 
3. If there is any additional calculation in the parse function, then only that part is overriden in the derived ExecParser and the rest of the function uses the GenereicExecParser base code.
4. Removed all the files which are not needed anymore
5. HashAggregateExec was getting incorrect metrics. Updated it to get the correct one.
6. Updated unit tests which had HashAggregateExec. Current unit tests in SqlPlanParserSuite passes with these changes.

Refactoring to use `GenericExecParser`:

* [`core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/CustomShuffleReaderExecParser.scala`](diffhunk://#diff-597ceccf94016cb6963289e9d36b91a710a4be183418dbe08acdeba8d5aa92b6L24-R30): Modified to extend `GenericExecParser` instead of implementing its own parsing logic.
* [`core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/GenericExecParser.scala`](diffhunk://#diff-171692b6f14db0908428557745327f788da85301dba7c16296a43ed493c7a2d9R1-R107): Added a new `GenericExecParser` class to handle common parsing logic for execution nodes.

### Removal of Individual Parser Classes:

* Removed multiple individual parser classes, including `AggregateInPandasExecParser`, `ArrowEvalPythonExecParser`, `CartesianProductExecParser`, `CoalesceExecParser`, `CollectLimitExecParser`, `ExpandExecParser`, `FilterExecParser`, `FlatMapGroupsInPandasExecParser`, `GenerateExecParser`, and others. These classes were responsible for parsing specific execution nodes but are now replaced by a more generic approach. [[1]](diffhunk://#diff-1593775a8f2daac6ac4323c2c722566dd1ed9b27020c9b4b8fb02ecad0de33a2L1-L41) [[2]](diffhunk://#diff-ed1884bb31b782326b5eae2c8b13b2d56eae92ebe14664edc89ceb000662633eL1-L41) [[3]](diffhunk://#diff-d4cb628102fc5d540771962bb8a56472fe9229cd484e481b407ae58bcea54de0L1-L41) [[4]](diffhunk://#diff-664ace0a91398a5511222d65d154346ec01900abd7ff8e7249e33a4216d39684L1-L42) [[5]](diffhunk://#diff-a471f51c1d1859aaf1f4a512f5f39e22efa2170a96d0425feded792dc56d41fcL1-L41) [[6]](diffhunk://#diff-1ddc4f93b00b25b1afa2ba01c63431ab46dd3d415b34f78f769576f006782ddaL1-L46) [[7]](diffhunk://#diff-68692022577ca087a560d1242f98203c2ccb661a89f305a4d1de26da21030cc0L1-L46) [[8]](diffhunk://#diff-634bbad80ee4a081660e53b11d832225aa2a669a89ea18266f48168d89b0ecfdL1-L41) [[9]](diffhunk://#diff-782d085ad887e316e8f37a0f5481517e738ec75808eff8ae73c2aebffb15c6d3L1-L46)

These changes streamline the parser configuration process, reduce redundancy, and make it easier to manage and extend the parser functionality.

tgravescs commented:
> Introduced json file which has all the information of Execs, expressions and so on. This can be extended to include other info such as join types, metrics and so on.

Can you please provide more details on the benefits of using a json file for this is better then just having it in the code? The base classes all make sense but having extra json file based on what I see implemented here seems like extra complexity that I want to understand.

nartal1 commented:
> > Introduced json file which has all the information of Execs, expressions and so on. This can be extended to include other info such as join types, metrics and so on.
> 
> Can you please provide more details on the benefits of using a json file for this is better then just having it in the code? The base classes all make sense but having extra json file based on what I see implemented here seems like extra complexity that I want to understand.

I added json file so that we have all the Execs information at one place and can add or update the Execs/expressions/configs easily. Also, if there are any fields to be added - like `joinType` we can update the json file instead of updating the code. 
Agree that it adds complexity of reading json and there is no type checking. The main reason was to decouple the parserNames and parserMethods from the code.
Would you prefer this information to be in the code itself?We can create a Map to hold all the Execs information something like below in ExecParserLoader.scala (instead of reading and parsing the JSON we could add it in the code itself):
```
val parsersConfig: Map[String, ParserConfig] = Map(
    "Union" -> ParserConfig(
      execName = "Union",
      parseExpressions = false,
    ),
    "TakeOrderedAndProject" -> ParserConfig(
      execName = "TakeOrderedAndProject",
      parseExpressions = true,
      expressionParserMethod = Some(sqlPlanParserType.decl(TermName("parseTakeOrderedExpressions")).asMethod),
      durationMetrics = None
```

tgravescs commented:
So its definitely an interesting solutions but without further information or reasons for this design we seem to be introducing more complexity, less readable, performance overhead, and losing type checking and binary compatibility checks for no reasons.

I honestly don't even know that you need a map. This can just be a case statement and some subclasses to reduce most of the code duplication.

amahussein commented:
> So its definitely an interesting solutions but without further information or reasons for this design we seem to be introducing more complexity, less readable, performance overhead, and losing type checking and binary compatibility checks for no reasons.
> 
> I honestly don't even know that you need a map. This can just be a case statement and some subclasses to reduce most of the code duplication.

There are some other reasons why would need to configure the tools' handling for different execs:
- In subtask of https://github.com/NVIDIA/spark-rapids-tools/issues/1116. Currently, we hardcode how each exec is assigned a category. This makes it difficult for:
   -  Internal teams won't know how an exec is going to be reported by the tools unless they run the tools cmd on an eventlog.
   - Switching categories of different execs require lots of code changes; especially with the overlap between shouldRemove/UDFs/special cases of ScanOps..etc.
   - Having execs configuration:
      - Makes it easy to spot that in one place
      - Allows internal team to experiment different the tools behavior by changing the configuration of specific execs.
      - Reduce the maintenance on the devs side since changes can be done by non-dev contributors.
- Adding new environments will be more straightforward without changing the core classes. For example, supporting Velox/Photon would be simplified by only implementing what's different for the parsing portion.
- Currently, adding new spark execs requires adding a new class which ends up being copy/paste from other existing classes.
  - It introduces bugs as we find that changes in one place needs to be carried over in ten other class files.
  - sync up with the plugin support files can be limited to update the configurations.
- Finally, one important reason to refactor the parsing code is to allow using that parsing from Profiler/Qualifcation context. Currently, the parser code assumes the qualification; however, the same functionality would be needed as more customer are going to run the Profiling tool (some users requested detailed operator reports in the profiling output)

tgravescs commented:
I think you might be missing my main point.  I'm not against refactoring it to make use of common classes. I mentioned this years ago. I'm questions the need for a separate json file and using reflection.

```
Internal teams won't know how an exec is going to be reported by the tools unless they run the tools cmd on an eventlog.
```

I don't know what this means, you have to run the tool to get the output.  how does this change magically report output?

```
Having execs configuration:

- Makes it easy to spot that in one place
- Allows internal team to experiment different the tools behavior by changing the configuration of specific execs.
- Reduce the maintenance on the devs side since changes can be done by non-dev contributors.

```

- it can be in one place in the code as well, and with a separate file I now have to have 2 files open to find what it actually is doing.  Harder to navigate via IDE.
- When would this experimentation happen? you either support an exec properly or you don't., The exec either supports expressions or it doesn't.  If you are adding it you either add a couple lines of code of couples lines in json file
- When would a non-dev modify this? This was going back to my question or if customers or someone is modifying it then this may make sense, but at the same time in my opinion this is also adding a lot of complexity.


```
Adding new environments will be more straightforward without changing the core classes. For example, supporting Velox/Photon would be simplified by only implementing what's different for the parsing portion.

```

I don't understand how this has anything to do with json/reflection vs doing it in code. Either case you are adding a few lines somewhere and do the testing.  I guess adding to a json file could be easier but you have the negatives already listed as well. A developer should be easily be able to add a couple lines of code, adding in code does all the proper type checking with less complexity.

```
 Currently, adding new spark execs requires adding a new class which ends up being copy/paste from other existing classes.
```

I already stated ways to remove the new class requirement. That was mostly done originally to parallelize development, other ways to reduce code without introducing json file and reflection.

```
Finally, one important reason to refactor the parsing code is to allow using that parsing from Profiler/Qualifcation context. Currently, the parser code assumes the qualification; however, the same functionality would be needed as more customer are going to run the Profiling tool (some users requested detailed operator reports in the profiling output)
```

This again doesn't require json and reflection to reuse code.  The question isn't about refactoring the code, the question is more about adding this json file and reflection.

nartal1 commented:
@tgravescs @amahussein - Please take another look and let me know if this approach seems okay.

tgravescs commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/CustomShuffleReaderExecParser.scala:
```diff
@@ -21,22 +21,11 @@ import com.nvidia.spark.rapids.tool.qualification.PluginTypeChecker
-    sqlID: Long) extends ExecParser {
+    override val node: SparkPlanGraphNode,
+    override val checker: PluginTypeChecker,
+    override val sqlID: Long) extends GenericExecParser(node, checker,sqlID) {
```
nit space before sqlID

tgravescs commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/CustomShuffleReaderExecParser.scala:
```diff
@@ -21,22 +21,11 @@ import com.nvidia.spark.rapids.tool.qualification.PluginTypeChecker
-    }
-    ExecInfo(node, sqlID, node.name, "", speedupFactor, duration, node.id, isSupported, None)
-  }
+  override val fullExecName = "CustomShuffleReaderExec"
```
Is this the only reason this isn't just a GenericExecParser?  Is it the only one or could we make the exec name a parameter to GenericExecParser and have it default to  node.name + "Exec"?

nartal1 replied:
Thanks for the tip! Removed this override and made exec name as a parameter.

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:
```diff
@@ -476,67 +476,67 @@ object SQLPlanParser extends Logging {
+        GenericExecParser(node, checker, sqlID, Some(parseWindowExpressions)).parse
       case "WindowInPandas" =>
-        WindowInPandasExecParser(node, checker, sqlID).parse
+        GenericExecParser(node, checker, sqlID).parse
```
Thanks @amahussein  for the suggestion! I thought of making this block as default after completing the refactor of all Execs. I will include the above change in this PR.

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/ObjectHashAggregateExecParser.scala:
```diff
@@ -23,29 +23,13 @@ import org.apache.spark.sql.execution.ui.SparkPlanGraphNode
+    override val checker: PluginTypeChecker,
+    override val sqlID: Long,
+    appParam: AppBase) extends
+  GenericExecParser(node, checker, sqlID) with Logging {
```
done

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:
```diff
@@ -464,9 +464,9 @@ object SQLPlanParser extends Logging {
     normalizedNodeName match {
       case "AggregateInPandas" =>
-        AggregateInPandasExecParser(node, checker, sqlID).parse
+        GenericExecParser(node, checker, sqlID).parse
```
done

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/GenericExecParser.scala:
```diff
@@ -0,0 +1,107 @@
+  // Compute duration based on the node metrics of that ExecNode
+  protected def computeDuration: Option[Long] = {
+    None
+  }
```
Thanks! Added another method to get the metrics.

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/GenericExecParser.scala:
```diff
@@ -0,0 +1,107 @@
+
+  val fullExecName: String = node.name + "Exec"
+
+  override def parse: ExecInfo = {
```
Keeping it as it is based on offline discussion. We can change it later if required.

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/planparser/GenericExecParser.scala:
```diff
@@ -0,0 +1,107 @@
+    None
+  }
+
+  protected def isSupportedByDefault: Boolean = true
```
My initial thought was to override this in the Exec where it is not supported by default. But we get that info already from isSupported function. Removing it for now.

nartal1 commented:
Thanks @amahussein and @tgravescs for the review. I have addressed review comments.
This PR doesn't include refactor of all the Execs but covers most of the ones which had common code.  PTAL.

amahussein reviewed (approved):
Thanks @nartal1 
LGTME!
